modules/unsupported/wfs-ng/src/main/java/org/geotools/data/wfs/impl/WFSContentComplexFeatureSource.java

```diff
@@ -99,12 +106,43 @@ public FeatureCollection<FeatureType, Feature> getFeatures(Query query) throws I
         request.setPropertyNames(query.getPropertyNames());
         request.setSortBy(query.getSortBy());
 
-        String srsName = null;
-        request.setSrsName(srsName);
+        if (query.getCoordinateSystem() != null) {
+            String srsName = findSupportedSrs(request, query.getCoordinateSystem());
+
+            if (srsName != null) {
+                request.setSrsName(srsName);
+            } else {
+                LOGGER.log(
+                        Level.WARNING,
+                        "WFS doesn't support the coordinate system: "
+                                + query.getCoordinateSystem());
+            }
+        }
 
         return new WFSContentComplexFeatureCollection(request, schema, name, client);
     }
 
+    private String findSupportedSrs(GetFeatureRequest request, CoordinateReferenceSystem crs) {
+        String identifier = GML2EncodingUtils.toURI(crs, SrsSyntax.AUTH_CODE, false);
+        if (identifier != null) {
+            int idx = identifier.lastIndexOf(':');
+            String authority = identifier.substring(0, idx);
+            String code = identifier.substring(idx + 1);
+            Set<String> supported =
+                    request.getStrategy().getSupportedCRSIdentifiers(request.getTypeName());
+            for (String supportedSrs : supported) {
+                if (supportedSrs.contains(authority) && supportedSrs.endsWith(":" + code)) {
+                    LOGGER.log(Level.FINE, "Found supportedSRS: " + supportedSrs);
+                    return supportedSrs;
+                }
+            }
+        } else {
+            LOGGER.log(
+                    Level.FINE, "GML2EncodingUtils couldn't handle the coordinate system: " + crs);
+        }
+        return null;
+    }
```

Comment:
Why replicate the same code found in WFSFeatureSource though? Better to share it in a static method.

Reply:
Instead of a static method, I made it a method of GetFeatureRequest. It's actually that classes srsName we want to set.